botocore/utils.py

```diff
@@ -794,5 +802,152 @@ def _switch_hosts(request, new_endpoint, use_new_scheme=True):
+                "in that region and the proper region could not be "
+                "automatically determined." % (client_region, bucket))
+            return
+
+        logger.debug(
+            "S3 client configured for region %s but the bucket %s is in region"
+            " %s; Please configure the proper region to avoid multiple "
+            "unnecessary redirects and signing attempts." % (
+                client_region, bucket, new_region))
+
+        endpoint = self._endpoint_resolver.resolve('s3', new_region)
+        endpoint = endpoint['endpoint_url']
+
+        signing_context = {
+            'region': new_region,
+            'bucket': bucket,
+            'endpoint': endpoint
+        }
+        request_dict['context']['signing'] = signing_context
+
+        self._cache[bucket] = signing_context
+        self.set_request_url(request_dict, request_dict['context'])
+
+        # Return 0 so it doesn't wait to retry
+        return 0
+
+    def get_bucket_region(self, response):
+        """
+        There are multiple potential sources for the new region to redirect to,
+        but they aren't all universally available for use. This will try to
+        find region from response elements, but will fall back to calling
+        HEAD on the bucket if all else fails.
+
+        :param response: A response representing a service request that failed
+            due to incorrect region configuration.
+        """
+        # First try to source the region from the headers.
+        service_response = response[1]
+        response_headers = service_response['ResponseMetadata']['HTTPHeaders']
+        if 'x-amz-bucket-region' in response_headers:
+            return response_headers['x-amz-bucket-region']
+
+        # Next, check the error body
+        region = service_response.get('Error', {}).get('Region')
+        if region:
+            return region
+
+        # Finally, HEAD the bucket. No other choice sadly.
+        bucket = service_response.get('Error', {}).get('Bucket')
+        try:
+            self._client.head_bucket(Bucket=bucket)
+            region = None
+        except ClientError as e:
+            headers = e.response['ResponseMetadata']['HTTPHeaders']
+            region = headers.get('x-amz-bucket-region', None)
+
+        return region
+
+    def set_request_url(self, params, context, **kwargs):
+        endpoint = context.get('signing', {}).get('endpoint', None)
+        if endpoint is not None:
+            params['url'] = _get_new_endpoint(params['url'], endpoint)
+
+    def redirect_from_cache(self, params, context, **kwargs):
+        """
+        This handler retrieves a given bucket's signing context from the cache
+        and adds it into the request context.
+        """
+        bucket = params.get('Bucket')
+        signing_context = self._cache.get(bucket)
+        if signing_context is not None:
+            context['signing'] = signing_context
+
+
+class ThreadSafeCache(MutableMapping):
```

Comment:
Looking through the code again, I do not see any race conditions. We should be fine with using a plain dict. I think I forgot that since dict is a built-in it is thread safe in nature or there might have been something that you were doing before that was not safe, but right now it would be fine to remove the lock.